File: entities/tank.py
"""坦克基类。"""
import pygame

import config
from entities.base import Entity
from entities.bullet import Bullet
from utils.helpers import load_image


class Tank(Entity):
    """玩家与敌人坦克的公共基类。"""

    def __init__(self, x, y, image_name, speed, shoot_cooldown):
        image = load_image(image_name, (config.TILE_SIZE, config.TILE_SIZE))
        super().__init__(x, y, image)
        self.direction = config.DIR_UP
        self.speed = speed
        self.shoot_cooldown = shoot_cooldown
        self.shoot_timer = 0.0
        self.active_bullet = None
        self.moving = False
        self._base_image = self.image

    def set_direction(self, direction):
        """设置坦克朝向并旋转图片。"""
        if self.direction == direction:
            return
        self.direction = direction
        angle = direction * 90
        center = self.rect.center
        self.image = pygame.transform.rotate(self._base_image, -angle)
        self.rect = self.image.get_rect(center=center)

    def can_move(self, dx, dy, level, other_tanks=None):
        """检查坦克是否可以沿指定增量移动。"""
        new_rect = self.rect.move(dx, dy)
        # 边界检查
        if new_rect.left < 0 or new_rect.right > config.MAP_WIDTH:
            return False
        if new_rect.top < 0 or new_rect.bottom > config.MAP_HEIGHT:
            return False
        # 瓦片碰撞
        if level.collides_with_rect(new_rect):
            return False
        # 与其他坦克碰撞
        if other_tanks:
            for tank in other_tanks:
                if tank is not self and tank.alive and new_rect.colliderect(tank.rect):
                    return False
        return True

    def try_move(self, dt, level, other_tanks=None):
        """尝试按当前方向移动。"""
        vx, vy = config.DIRECTION_VECTORS[self.direction]
        dx = vx * self.speed * dt
        dy = vy * self.speed * dt
        # 先按原始方向移动；如果不行，尝试分别沿 x/y 轴滑动
        if self.can_move(dx, dy, level, other_tanks):
            self.rect.x += dx
            self.rect.y += dy
            return True
        if dx != 0 and self.can_move(dx, 0, level, other_tanks):
            self.rect.x += dx
            return True
        if dy != 0 and self.can_move(0, dy, level, other_tanks):
            self.rect.y += dy
            return True
        return False

    def shoot(self, owner):
        """发射子弹。"""
        if self.shoot_timer > 0:
            return None
        if self.active_bullet is not None and self.active_bullet.alive:
            return None
        self.shoot_timer = self.shoot_cooldown
        vx, vy = config.DIRECTION_VECTORS[self.direction]
        cx, cy = self.rect.center
        bullet_size = config.TILE_SIZE // 4
        # 子弹从坦克炮口射出，初始位置在坦克外边缘外 2 像素，避免自碰撞
        offset = config.TILE_SIZE // 2 + bullet_size // 2 + 2
        bullet_x = cx - bullet_size // 2 + vx * offset
        bullet_y = cy - bullet_size // 2 + vy * offset
        bullet = Bullet(bullet_x, bullet_y, self.direction, owner, self)
        self.active_bullet = bullet
        return bullet

    def update(self, dt):
        """更新射击冷却。"""
        if self.shoot_timer > 0:
            self.shoot_timer = max(0.0, self.shoot_timer - dt)
        if self.active_bullet is not None and not self.active_bullet.alive:
            self.active_bullet = None

    def take_damage(self):
        """受击处理，子类重写。"""
        self.kill()

    def respawn(self, x, y):
        """复活/重置位置。"""
        self.rect.topleft = (x, y)
        self.direction = config.DIR_UP
        self.image = self._base_image
        self.alive = True
        self.shoot_timer = 0.0
        self.active_bullet = None

File: js/entities/tank.js
"use strict";

var CONFIG = require('../config');
var Entity = require('./base');
var Bullet = require('./bullet');
var helpers = require('../utils/helpers');

function Tank(x, y, asset, speed, shootCooldown) {
  Entity.call(this, x, y, CONFIG.TILE_SIZE, CONFIG.TILE_SIZE);
  this.asset = asset;
  this.direction = CONFIG.DIR_UP;
  this.speed = speed;
  this.shootCooldown = shootCooldown;
  this.shootTimer = 0;
  this.activeBullet = null;
  this.moving = false;
}

Tank.prototype = Object.create(Entity.prototype);
Tank.prototype.constructor = Tank;

Tank.prototype.setDirection = function (direction) {
  if (this.direction === direction) {
    return;
  }
  this.direction = direction;
};

Tank.prototype.canMove = function (dx, dy, level, otherTanks) {
  var newRect = {
    x: this.x + dx,
    y: this.y + dy,
    w: this.w,
    h: this.h
  };

  if (newRect.x < 0 || newRect.x + newRect.w > CONFIG.MAP_WIDTH) {
    return false;
  }
  if (newRect.y < 0 || newRect.y + newRect.h > CONFIG.MAP_HEIGHT) {
    return false;
  }
  if (level.collidesWithRect(newRect)) {
    return false;
  }

  if (otherTanks) {
    for (var i = 0; i < otherTanks.length; i++) {
      var tank = otherTanks[i];
      if (tank !== this && tank.alive && helpers.rectsIntersect(newRect, tank.getRect())) {
        return false;
      }
    }
  }
  return true;
};

Tank.prototype.tryMove = function (dt, level, otherTanks) {
  var vec = CONFIG.DIRECTION_VECTORS[this.direction];
  var vx = vec[0];
  var vy = vec[1];
  var dx = vx * this.speed * dt;
  var dy = vy * this.speed * dt;

  if (this.canMove(dx, dy, level, otherTanks)) {
    this.x += dx;
    this.y += dy;
    return true;
  }
  if (dx !== 0 && this.canMove(dx, 0, level, otherTanks)) {
    this.x += dx;
    return true;
  }
  if (dy !== 0 && this.canMove(0, dy, level, otherTanks)) {
    this.y += dy;
    return true;
  }
  return false;
};

Tank.prototype.shoot = function (owner) {
  if (this.shootTimer > 0) {
    return null;
  }
  if (this.activeBullet && this.activeBullet.alive) {
    return null;
  }
  this.shootTimer = this.shootCooldown;

  var vec = CONFIG.DIRECTION_VECTORS[this.direction];
  var vx = vec[0];
  var vy = vec[1];
  var cx = this.x + this.w / 2;
  var cy = this.y + this.h / 2;
  var bulletSize = Math.floor(CONFIG.TILE_SIZE / 4);
  var offset = Math.floor(CONFIG.TILE_SIZE / 2 + bulletSize / 2 + 2);
  var bx = cx - bulletSize / 2 + vx * offset;
  var by = cy - bulletSize / 2 + vy * offset;

  var bullet = new Bullet(bx, by, this.direction, owner, this);
  this.activeBullet = bullet;
  return bullet;
};

Tank.prototype.update = function (dt) {
  if (this.shootTimer > 0) {
    this.shootTimer = Math.max(0, this.shootTimer - dt);
  }
  if (this.activeBullet && !this.activeBullet.alive) {
    this.activeBullet = null;
  }
};

Tank.prototype.takeDamage = function () {
  this.kill();
};

Tank.prototype.respawn = function (x, y) {
  this.x = x;
  this.y = y;
  this.direction = CONFIG.DIR_UP;
  this.alive = true;
  this.shootTimer = 0;
  this.activeBullet = null;
};

Tank.prototype.draw = function (ctx) {
  var cx = this.x + this.w / 2;
  var cy = this.y + this.h / 2;
  ctx.save();
  ctx.translate(cx, cy);
  ctx.rotate(this.direction * Math.PI / 2);
  this.asset(ctx, -this.w / 2, -this.h / 2, this.w);
  ctx.restore();
};

module.exports = Tank;
